aquadoggo/src/db/stores/task.rs

```diff
@@ -0,0 +1,210 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use anyhow::Result;
+use p2panda_rs::document::{DocumentId, DocumentViewId};
+use sqlx::{query, query_as};
+
+use crate::db::errors::SqlStorageError;
+use crate::db::models::TaskRow;
+use crate::db::provider::SqlStorage;
+use crate::materializer::{Task, TaskInput};
+
+/// Methods to interact with the `tasks` table in the database.
+impl SqlStorage {
+    /// Inserts a "pending" task into the database.
+    pub async fn insert_task(&self, task: &Task<TaskInput>) -> Result<(), SqlStorageError> {
+        // Convert task input to correct database types
+        let task_input = task.input();
+        let document_id = task_input.document_id.as_ref().map(|id| id.as_str());
+        let document_view_id = task_input
+            .document_view_id
+            .as_ref()
+            .map(|view_id| view_id.as_str());
+
+        // Check first if this task already exists, to avoid duplicate rows
+        let task_row = query_as::<_, TaskRow>(
+            "
+            SELECT
+                name,
+                document_id,
+                document_view_id
+            FROM
+                tasks
+            ",
+        )
+        .fetch_optional(&self.pool)
+        .await
+        .map_err(|err| SqlStorageError::Transaction(err.to_string()))?;
+
+        // If yes, we are already done here
+        if task_row.is_some() {
+            return Ok(());
+        }
```

Comment:
We could spare this query with a unique constraint on the tasks table couldn't we?

Reply:
Yes, No :sweat_smile: When the pending tasks are loaded into the queue during rescheduling, they are not being removed from the database (yet). The worker itself will fire `on_update` with a new `TaskStatus::Pending` as soon as it got queued, which leads to a new row in the `tasks` table and ultimately to a duplicate.

The worker itself makes sure that no duplicates exist in the queues, but thats not the case with the database.

I guess one could take the rows out of the database _after_ we run  `get_tasks` and _before_ we queue them up. Or we need a new queing approach which is used for rescheduling tasks which should lead to not firing `TaskStatus::Pending`? Or of course, we just check before insertion if there is already something as it is now. :thinking: